pkg/auth/oidc_connect.go

```diff
@@ -0,0 +1,294 @@
+	}
+
+	// 2. Check do we have stateid stored in querystring ?
+	var state *store.OIDCState
+	validState := false
+
+	stateToken := url.Query().Get(stateQueryParamName)
+	stateByte, err := h.Cache.Get(stateToken)
+	if err == nil {
+		state = store.ConvertToType(stateByte)
+	}
+
+	// 2.1 State not found, try to retrieve from cookies.
+	if state == nil {
+		state, _ = h.getStateFromCookie(ctx, req)
+	}
+
+	//2.2 State exist, proceed with token validation.
+	if state != nil {
+		// Re-initialize provider to refresh the context, this seems like bugs with coreos go-oidc module.
+		provider, err := h.initProvider(ctx)
+		if err != nil {
+			h.Log.Info(fmt.Sprintf("fail to initialize provider: %v", err))
+			return
+		}
+		validState = h.isStateValid(ctx, state, provider)
+
+	}
+
+	if validState {
+		stateJSON, _ := json.Marshal(state)
+		// Restore cookies.
+		resp.Response.Header.Add(oauthTokenName, string(stateJSON))
+		h.Cache.Delete(state.OAuthState)
+		resp.Allow = true
+
+	} else {
+		// 2.3 Route to login handler.
+		h.loginHandler(req, resp, url)
+	}
+
+}
+
+func (h *OIDCConnect) loginHandler(req *Request, resp *Response, u *url.URL) {
+
+	state := store.NewState()
+	state.GenerateOauthState()
+	state.RequestPath = u.Host + u.Path
+
+	resp.Response.Header.Add(oauthTokenName, "")
+
+	authCodeURL := h.oauth2Config().AuthCodeURL(state.OAuthState)
+	byteState := store.ConvertToByte(state)
+
+	h.Cache.Set(state.OAuthState, byteState)
+	resp.Response.StatusCode = http.StatusFound
+	resp.Response.Header.Add("Location", authCodeURL)
+}
+
+func (h *OIDCConnect) callbackHandler(ctx context.Context, req *Request, resp *Response, u *url.URL) {
+
+	// 1. Get all variable needed.
+	oauthState := u.Query().Get("state")
+	code := u.Query().Get("code")
+
+	// Check state and code validity.
+	if code == "" || oauthState == "" {
+		resp.Response.StatusCode = http.StatusBadRequest
+		resp.Allow = false
+		return
+	}
+
+	stateByte, err := h.Cache.Get(oauthState)
+	if err != nil {
+		h.Log.Info("Invalid state, expected " + oauthState + " not found ")
+		resp.Response.StatusCode = http.StatusBadRequest
+		resp.Allow = false
+		return
+	}
+
+	// Retrieve token.
+	context := oidc.ClientContext(ctx, h.HTTPClient)
+	token, err := h.oauth2Config().Exchange(context, code)
+	if err != nil {
+		h.Log.Info(err.Error())
+	}
+
+	rawIDToken, ok := token.Extra("id_token").(string)
+
+	if !ok {
+		resp.Response.StatusCode = http.StatusInternalServerError
+		resp.Allow = false
+		return
+	}
+
+	//Store token.
+	state := store.ConvertToType(stateByte)
+	state.IDToken = rawIDToken
+	state.AccessToken = token.AccessToken
+	state.RefreshToken = token.RefreshToken
+	stateByte = store.ConvertToByte(state)
+	h.Cache.Set(state.OAuthState, stateByte)
+
+	// Set  redirection.
+	resp.Response.StatusCode = http.StatusTemporaryRedirect
+	resp.Response.Header.Add(stateQueryParamName, state.OAuthState)
+
+	// TODO .. may need to rebuild this, in case user could have other query parameters
```

Comment:
Yeh this makes sense. Once we land this, let's file issues for any outstanding TODO items.